use importlib for static resources

behrmann opened the pull request:
Now that we are on python 3.7, we can finally use [`importlib.resources`](https://docs.python.org/3.7/library/importlib.html#module-importlib.resources) and move static content into a submodule, which gives us proper editor support for those files, while making the main `__init__.py` a couple hundred lines shorter.

This could probably be combined with loading formatable strings from there and running `.format` on them later, but I haven't tried that yet, because I wanted to check first, whether we want to do this.

DaanDeMeyer commented:
I had no idea this was possible. Neat!

keszybz commented on mkosi/__init__.py:
```diff
@@ -2621,161 +2552,35 @@ def install_arch(args: CommandLineArguments, root: Path, do_run_build_script: bo
-                Exec = /etc/pacman.d/scripts/mkosi-kernel-add
-                NeedsTargets
-                """
-            )
+            importlib.resources.read_text("mkosi.resources.arch", "60_kernel_remove.hook")
         )
```
I think it'd make sense to add a helper for this:
```py
def write_resource(where: Path, resource: str, key: str, bool: executabe = False) -> None:
    text = importlib.resources.read_text(resource, key)
    where.write_text(text)
    if executable: …

write_resource(hooks_dir / '90-mkosi-kernel-add.hook', 'mkosi.resources.arch', '60_kernel_remove.hook')
```

```

behrmann replied:
Will change it accordingly.

behrmann replied:
Added a `mode` switch as well.

behrmann commented:
> EDIT: Actually, how does this work? does this mean we need to install all the resources alongside the python files? I don't see anything integrating this with setup.py on first sight so I'm curious how this would work.

Good question, I'm still in the process of figuring this out. :) So far I've only used this feature together with zipapps.

keszybz reviewed (commented):
This is great.

behrmann commented:
Ok, now it's hooked up to `setup.py` and the files are installed alongside accordingly.

While at it I removed the executable bit from two hook files in Arch, because I don't think they need to be executable and switched some `os.makedirs` to the pathlib calls.

behrmann commented:
Just changed the explicit list of packages in `setup.py` for setuptools' `find_packages` to be futureproof.

keszybz commented on mkosi/__init__.py:
```diff
@@ -2551,37 +2565,32 @@ def install_arch(args: CommandLineArguments, root: Path, do_run_build_script: bo
+            "mkosi.resources.arch",
+            "kernel_remove.sh",
+            executable=True,
+        )
```
This really should be formatted as 
```py
       write_resource(hooks_dir / "60-mkosi-kernel-remove.hook", "mkosi.resources.arch", "60_kernel_remove.hook")
       write_resource(scripts_dir / "mkosi-kernel-remove", "mkosi.resources.arch", "kernel_remove.sh",
                                       executable=True)
```
Black makes it seem like two completely diferent things. OTOH, the PEP-8-style approach makes it easy to do a spot differences and similarities.

Obviously the same applies below: for no good reason, half of the calls to write_resource are compact, and half are spread.

behrmann replied:
I'm not quite sold, that the PEP-8 style makes it easier to spot differences, but I'm generally not fond of overly long lines and prefer the more vertical style, as I find it more declarative. But this is something we can argue for a long time, now it's in the style my Emacs produces without interference :)

keszybz reviewed (commented):
LGTM. I dislike the formatting, but I understand that this matches your preference, so I won't argue.

DaanDeMeyer commented:
All these install tests you wrote a long time ago will come in handy now. If everything passes this is good to merge.

poettering commented:
Amazeballs!
